core/src/test/java/kafka/server/integration/EligibleLeaderReplicasIntegrationTest.java

```diff
@@ -0,0 +1,432 @@
+        Map<ConfigResource, Collection<AlterConfigOp>> configOps = Collections.singletonMap(configResource, ops);
+        // alter configs on target cluster
+        adminClient.incrementalAlterConfigs(configOps).all().get();
+        Producer producer = null;
+        Consumer consumer = null;
+        try {
+            // check which partition is on broker 0 which we'll kill
+            TopicDescription testTopicDescription = adminClient.describeTopics(Collections.singletonList(testTopicName))
+                    .allTopicNames().get().get(testTopicName);
+            TopicPartitionInfo topicPartitionInfo = testTopicDescription.partitions().get(0);
+            List<Node> initialReplicas = topicPartitionInfo.replicas();
+            assertEquals(4, topicPartitionInfo.isr().size());
+            assertEquals(0, topicPartitionInfo.elr().size());
+            assertEquals(0, topicPartitionInfo.lastKnownElr().size());
+
+            Properties producerProps = new Properties();
+            producerProps.putIfAbsent(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+            producerProps.putIfAbsent(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+            producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer);
+            producerProps.put(ProducerConfig.ACKS_CONFIG, "1");
+            producer = new KafkaProducer(producerProps);
+
+            Properties consumerProps = new Properties();
+            consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer);
+            consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "test");
+            consumerProps.put(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, "0");
+            consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+            consumerProps.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+            consumerProps.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+            consumer = new KafkaConsumer<>(consumerProps);
+            consumer.subscribe(Collections.singleton(testTopicName));
+
+            producer.send(new ProducerRecord<>(testTopicName, "0", "0")).get();
+            Thread.sleep(1000);
+            ConsumerRecords records = consumer.poll(Duration.ofSeconds(1L));
+            assertEquals(1, records.count());
+
+            killBroker(initialReplicas.get(0).id());
+            killBroker(initialReplicas.get(1).id());
+
+            waitForIsrAndElr((isrSize, elrSize) -> {
+                return isrSize == 2 && elrSize == 1;
+            });
+
+            // Now the partition is under min ISR. HWM should not advance.
+            producer.send(new ProducerRecord<>(testTopicName, "1", "1")).get();
+            Thread.sleep(1000);
+            records = consumer.poll(Duration.ofSeconds(1L));
+            assertEquals(0, records.count());
```

Comment:
I wonder if we should just check that produce with acks=all fails with the not enough replicas error.  Trying to consume the records adds time to the test run (the sleep is real time not mock time) and doesn't guarantee that we didn't consume the record yet because HWM didn't move and not for some other reason (e.g. cluster is just slow).

Reply:
It is a valid concern. But I don't have an idea how to get around this. 
With ack=all, the producer request will be rejected. Then we don't know if restoring the min ISR can unblock the HWM with no offsets to be forwarded.